tests/docker_push_latest_if_changed_test.py

```diff
@@ -1,8 +1,463 @@
+
+
+@pytest.fixture
+def mock_tag_image():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_tag_image',
+        autospec=True,
+    ) as _mock_tag_image:
+        yield _mock_tag_image
+
+
+@pytest.fixture
+def mock_push_image():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_push_image',
+        autospec=True,
+    ) as _mock_push_image:
+        yield _mock_push_image
+
+
+@pytest.fixture
+def mock_is_image_changed():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_is_image_changed',
+        autospec=True,
+    ) as _mock_is_image_changed:
+        yield _mock_is_image_changed
+
+
+@pytest.fixture
+def mock_validate_source():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_validate_source',
+        autospec=True,
+    ) as _mock_validate_source:
+        yield _mock_validate_source
+
+
+@pytest.fixture
+def mock_get_image_parity_fields():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_get_image_parity_fields',
+        autospec=True,
+    ) as _mock_get_image_parity_fields:
+        yield _mock_get_image_parity_fields
+
+
+@pytest.fixture
+def mock_get_image_cmds_hash():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_get_image_cmds_hash',
+        autospec=True,
+    ) as _mock_get_image_cmds_hash:
+        yield _mock_get_image_cmds_hash
+
+
+@pytest.fixture
+def mock_get_packages_hash():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_get_packages_hash',
+        autospec=True,
+    ) as _mock_get_packages_hash:
+        yield _mock_get_packages_hash
+
+
+@pytest.fixture
+def mock_get_docker_cmds_blob():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_get_docker_cmds_blob',
+        autospec=True,
+    ) as _mock_get_docker_cmds_blob:
+        yield _mock_get_docker_cmds_blob
+
+
+@pytest.fixture
+def mock_get_sha256_hexdigest():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_get_sha256_hexdigest',
+        autospec=True,
+    ) as _mock_get_sha256_hexdigest:
+        yield _mock_get_sha256_hexdigest
+
+
+@pytest.fixture
+def mock_sha256():
+    _mock_sha256 = mock.Mock()
+    with mock.patch(
+        'hashlib.sha256',
+        autospec=True,
+        return_value=_mock_sha256
+    ):
+        yield _mock_sha256
+
+
+@pytest.fixture
+def mock_docker_client_creation():
+    with mock.patch.object(
+        docker,
+        'Client',
+        autospec=True
+    ) as _mock_docker_client_creation:
+        yield _mock_docker_client_creation
+
+
+@pytest.fixture
+def mock_docker_push_latest_if_changed():
+    with mock.patch.object(
+        docker_push_latest_if_changed,
+        '_docker_push_latest_if_changed',
+        autospec=True
+    ) as _mock_docker_push_latest_if_changed:
+        yield _mock_docker_push_latest_if_changed
```

Comment:
sorry but none of these tests seem actually useful.  You're testing _implementation_ and not behaviour.

One could implement this in a slightly different way and your entire testsuite would fail but it would be perfectly functional.

I'd suggest writing a few tests which set up some real images, a fake registry, and ensuring this works end-to-end.

Reply:
`capsys`, `tmpdir`, and a few `yield` fixtures can help you a lot in accomplishing this :)

Reply:
I stuck with unit tests since I didn't think you could test with something "third-party" ish like docker on travis CI. Did a bit more research after your commend and learned that you could.

Switched to doing end-to-end integration tests that run with docker and  fake registry. It's a lot slower, but should also give a lot more confidence. I agree that the previous tests weren't especially meaningful as they were just unit tests which kind of inherently only test implementation as you said.

Reply:
<3 thanks a million this :D